### PR TITLE
Remove admin session and add permissions API

### PR DESCRIPTION
**Pull request by michaelmattig**

### CODE REVIEW

**Comment by michaelmattig** [comment]
[Admin kann Rechte vergeben](https://trello.com/c/3H9dWVE0/1107-admin-kann-rechte-vergeben)

File: geoengine/permissions.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+    headers = session.auth_header
+
+    response = req.put(
+        url=f'{session.server_url}/roles',
+        headers=headers,
+        timeout=timeout,
+        json={
+            "name": name
+        }
```
Comment (jdroenner) [inline geoengine/permissions.py:148]: there should be a type for that in api.py

File: geoengine/auth.py [hunk 2]
```diff
@@ -206,8 +173,7 @@ def get_session() -> Session:
 
 def initialize(server_url: str,
                credentials: Optional[Tuple[str, str]] = None,
-               token: Optional[str] = None,
-               admin_token: Optional[str] = None) -> None:
+               token: Optional[str] = None) -> None:
```
Comment (jdroenner) [inline geoengine/auth.py:176]: so we can still use the token?

File: geoengine/permissions.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+from __future__ import annotations
+
+from typing import Dict
+from uuid import UUID
+import json
+
+import requests as req
+
+from geoengine.api import Permission, ResourceId
```
Comment (jdroenner) [inline geoengine/permissions.py:13]: please don't re-export types from api. It should be clear that they are only API types. We might generate them one day and then everything breaks.

File: geoengine/__init__.py [hunk 4]
```diff
@@ -11,10 +11,12 @@
 from .error import GeoEngineException, InputException, UninitializedException, TypeException, \
     MethodNotCalledOnPlotException, MethodNotCalledOnRasterException, MethodNotCalledOnVectorException, \
     SpatialReferenceMismatchException, check_response_for_error, ModificationNotOnLayerDbException, \
-    NoAdminSessionException, InvalidUrlException, MissingFieldInResponseException
+    InvalidUrlException, MissingFieldInResponseException
 from .layers import Layer, LayerCollection, LayerListing, LayerCollectionListing, \
     LayerId, LayerCollectionId, LayerProviderId, \
     layer_collection, layer
+from .permissions import add_permission, remove_permission, add_role, remove_role, assign_role, revoke_role, \
+    ADMIN_ROLE_ID, REGISTERED_USER_ROLE_ID, ANONYMOUS_USER_ROLE_ID, ResourceId, Permission, UserId, RoleId
```
Comment (jdroenner) [inline geoengine/__init__.py:19]: either remove the re-export or add them as real types. Would be nice to somehow get a ResourceId from a LayerId without writing the dict. Maybe add a from_layer_id(layer_id) to the new type. Or add an "ressource_id(self)" to LayerId

File: geoengine/permissions.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+REGISTERED_USER_ROLE_ID: RoleId = RoleId(UUID("4e8081b6-8aa6-4275-af0c-2fa2da557d28"))
+ANONYMOUS_USER_ROLE_ID: RoleId = RoleId(UUID("fd8e87bf-515c-4f36-8da6-1a53702ff102"))
+
+
+def add_permission(role: RoleId, resource: ResourceId, permission: Permission, timeout: int = 60):
+    """Add a permission to a resource for a role. Requires admin role."""
+
+    session = get_session()
+
+    payload = json.dumps({
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply geoengine/permissions.py]: cb77f6e92122f0cc346dbc4ab2320a44a0ceea55

File: geoengine/permissions.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        headers=headers,
+        timeout=timeout,
+        data=payload
+    )
+
+    if not response.ok:
+        raise GeoEngineException(response.json())
+
+
+def remove_permission(role: UUID, resource: ResourceId, permission: Permission, timeout: int = 60):
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply geoengine/permissions.py]: cb77f6e92122f0cc346dbc4ab2320a44a0ceea55

File: geoengine/permissions.py (hunk 1 above)
Comment (michaelmattig) [reply geoengine/permissions.py]: cb77f6e92122f0cc346dbc4ab2320a44a0ceea55

File: geoengine/auth.py (hunk 2 above)
Comment (michaelmattig) [reply geoengine/auth.py:176]: sure, the token is just the session id which is used in the authorization header

File: geoengine/permissions.py (hunk 3 above)
Comment (michaelmattig) [reply geoengine/permissions.py]: cb77f6e92122f0cc346dbc4ab2320a44a0ceea55

File: geoengine/__init__.py (hunk 4 above)
Comment (michaelmattig) [reply geoengine/__init__.py]: cb77f6e92122f0cc346dbc4ab2320a44a0ceea55

File: geoengine/permissions.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        return ResourceId('LayerCollection', str(layer_collection_id))
+
+    @classmethod
+    def from_dataset_id(cls, dataset_id: DatasetId) -> ResourceId:
+        '''Create a resource id from a dataset id'''
+        return ResourceId('Dataset', str(dataset_id))
+
+    def to_api_dict(self) -> api.ResourceId:
+        '''Convert to a dict for the API'''
+        return {
```
Comment (jdroenner) [inline geoengine/permissions.py:110]: return api.ResourceId()

**Review by jdroenner — CHANGES_REQUESTED** [review]

File: geoengine/permissions.py (hunk 7 above)
Comment (michaelmattig) [reply geoengine/permissions.py]: why does neither pylint nor mypy catch that?

8d71fe167e49c4484dd2deda9733b5e02562d47f

Comment (jdroenner) [reply geoengine/permissions.py]: i dont know :(

**Review by jdroenner — APPROVED** [review]

**Comment by jdroenner** [comment]
bors merge

**Pull request merged**